### ISSUE TITLE
Name clash between wildcard import and class in same package regression 2.12 -> 2.13

### ISSUE THREAD

**Issue by oxbowlakes** [scala/bug#11593]
Consider the following code in the source root `src/main/scala/foo/X.scala`:

    package foo {
      class Properties

      import java.util._
      object X extends App {
        def bar(x: Properties): Unit = println(x.getClass.getName)
        bar(new Properties)
      }
    }

It's my assumption that this code is equivalent to having the following

In `src/main/scala/foo/Properties.scala`

    package foo
    class Properties

In `src/main/scala/foo/X.scala`

    package foo
    import java.util._
    object X extends App {
      def bar(x: Properties): Unit = println(x.getClass.getName)
      bar(new Properties)
    }

In scala 2.12 the two seem to be equivalent; in both cases the program X prints `foo.Properties`. However in scala 2.13 this is no longer the case. The former still prints `foo.Properties` but the latter prints `java.util.Properties`

So I think that there are two separate issues here:

 1. That in scala 2.13 the two separate ways of constructing the `foo` package are not equivalent
 2. That there is a regression in the behaviour of the second (and probably more typical) construction. That a reference to `Properties` in the body of `X` would previously (in 2.12.x) have been resolved against `foo.Properties` but is now (in 2.13.x) resolved against `java.util.Properties`

**Comment by odersky** [comment]
In fact, this fixes a previous regression (I am not sure from when on). The Scala 2.13 behavior is according to spec. Here it is in [Chapter 2](https://scala-lang.org/files/archive/spec/2.12/02-identifiers-names-and-scopes.html):

Bindings of different kinds have a precedence defined on them:

 - Definitions and declarations that are local, inherited, or made available by a package clause and also defined in the same compilation unit as the reference, have highest precedence.
 - Explicit imports have next highest precedence.
 - Wildcard imports have next highest precedence.
 - Definitions made available by a package clause, but not also defined in the same compilation unit as the reference, have lowest precedence.

The reason for the spec wording is that one often has little insight or control over what is in a package. E.g. anything could be dumped in the empty package. So it is dangerous and unintuitive that this should shadow an import.

**Comment by dwijnand** [comment]
While acknowledging this is as specified, the thing I find troubling is the fact that the behaviour is different only depending on if it's in the same compilation unit or not.  To me that feels like it's breaking some kind of "law" or invariant.

Are there other examples where such a difference is present?

**Comment by odersky** [comment]
`sealed` also depends on the compilation unit.

**Comment by dwijnand** [comment]
Sure, you must use the same compilation unit to have access to and benefit from `sealed`-ness.

But that's different, here the same definitions split into separate files changes the behaviour.

**Comment by som-snytt** [comment]
Companionship also depends on the compilation unit.

The law is "what you see is what you get". The invisible thing that is available doesn't shadow what you read from the source text, which asks for an import explicitly.

This is more compelling for nested packages, where the available thing is in a sense more distant, not in the immediately enclosing package but higher up.

**Comment by dwijnand** [comment]
> Companionship also depends on the compilation unit.

That's a good example.  Companionship (e.g. with type aliases) can be a pitfall unless you know all the intricacies.

I don't understand your nested package example, but I'm fine with the argumentation being that `import java.util._` should mean that `java.util.Properties` is printed.  My argument is that it should be consistent, i.e. also print that in the single file example (example 1).

**Comment by oxbowlakes** [comment]
Apologies for not finding this (despite searching to see if it was an intended change)

**Comment by som-snytt** [comment]
I think https://github.com/scala/bug/issues/9552 is a good example and the `pos/t9552` test. Do you see a vase or two faces?

Appropriately, facing braces look like the famous illusion: `} {`

The precedence rules just say what can shadow something else, at the same nesting level or not the same nesting level.

In the example, moving the import inside the object renders it ambiguous.

The nested package example:
```
package a   // makes a.A visible
package b
import c.A
// which A do you expect? Note that a.A did not exist when I wrote the import.
// adding a.A shouldn't break this code.
```

**Comment by dwijnand** [comment]
I'm saying `c.A` sounds fine, whether `a.A` is defined in this file or any other, at least it's consistent behaviour.

I guess I'll take my argument to the contributors ML.

**Comment by oxbowlakes** [comment]
What is perhaps interesting is that this one change (fix) has caused us _by far_ the most pain in project migration to 2.13 because the compiler errors look like "method foo does not exist on Bar" and IDEs have not necessarily been "made aware" of the change (i.e. the IDE thinks the code is fine, only scalac thinks it's broken)

What is more, the wildcard import makes it often extremely difficult to understand where the other class could be coming from!

Note also that the argument provided by @som-snytt is not very convincing:

    package a //let's say that a.A exists
    import foo._
    new A //let's say that foo.A is later added. My code is now broken

**Comment by som-snytt** [comment]
There were a couple of Scala tests and a few cases in the community build where imports were wrongly used, so it's surprising to learn that this beats everything else as a pain point. Sorry to hear, of course.

I've intended to write a linter for lexical and implicit scopes, to explain shadowing errors; mostly for implicits, and also to explain where desired implicit types can be found; but it would be useful for any incoherent namespace where there's lots of globbing. Where does anything come from?

I think I would stand by my assertion that `import foo._` looks more specific than the package statement, although both are effectively globs.

**Comment by SethTisue** [comment]
PR was https://github.com/scala/scala/pull/6589

**Comment by SethTisue** [comment]
> the argument provided by @som-snytt is not very convincing

I find it convincing. @som's example does not involve a wildcard import. Your example does.

**Comment by dwijnand** [comment]
Here's a slightly modified version of @oxbowlakes's example where you take two files, concatenate them, and now compiling and running that program has different behaviour:

```scala
$ cat file1.scala
package foo { class Properties }
$ cat file2.scala
package foo { import java.util._; object X extends App { println((new Properties).getClass) } }
$ scalac file1.scala file2.scala && scala foo.X
class java.util.Properties
$ cat file1.scala file2.scala > file.scala
$ rm -r foo
$ scalac file.scala && scala foo.X
class foo.Properties
```

Does that seem right?

**Comment by SethTisue** [comment]
> Does that seem right?

Isn't it inherent to the combination of 1) separate compilation and 2) packages being open?

**Comment by som-snytt** [comment]
It comes down to whether you think the concept of "compilation unit" is significant.

Note that "compilation unit" usually means "file" but not always. In the REPL, a line is a compilation unit. Java has always highlighted that classes can come from anywhere.

Several years ago, I also thought it was strange that compilation units matter (for a few things), but now I have embraced the distinction.

**Comment by dwijnand** [comment]
> Isn't it inherent to the combination of 1) separate compilation and 2) packages being open?

Why can't we take those factors and then make it have the same behaviour in both cases?

I don't care if you want to argue that "import java.util._ comes after so it makes sense it's java.util.Properties" or "you're in package foo, so despite the import it makes sense it's foo.Properties".  I just care that it does the same thing no matter if I do the same thing in two files or the same thing in one file.

**Comment by som-snytt** [comment]
>  I just care that it does the same thing no matter if I do the same thing in two files or the same thing in one file.

I don't mean to cut off discussion on this point, but to me, it's like saying, I want this to work the same whether the code is in the same package or different packages.

Per my previous comment, either one is committed to "compilation unit" as a thing, or one really means to say this concept is void.

Note that Scala is a line-oriented language, and even blank lines are semantically significant.

Of course, people also complain, "I added a blank line and it changed the meaning."

**Comment by som-snytt** [comment]
Actually it's worth an `-Xlint`. I bet the current code can detect Dale's case without a lot of effort.

**Comment by dwijnand** [comment]
I accept a compilation unit is a thing.  But I think it should operate at the "you can't extend this because it's sealed and in that other file" level, and not in such a subtle way.  Companion objects are similarly problematic (and I don't have a solution there).

**Comment by SethTisue** [comment]
@dwijnand now that I take a closer look at your example, it does seem wrong to me that the shadowing behavior would be different in those two cases. I hadn't completely understood this before.

@som-snytt I do accept that compilation units are a thing, but I don't understand why being in the same compilation unit or not is, or should be, relevant in this case...? shouldn't the import always win?

**Comment by som-snytt** [comment]
I'll contribute an explanation in a doc page, and then folks can complain about that.

@SethTisue Yes, the import wins versus an arbitrary definition in an open package.

There's a whole resurgence in gothic cinema such as Get Out and also a nytimes piece on novels about motherhood as gothic horror ([What to expect when you're expecting evil](https://www.nytimes.com/2019/07/06/books/horror-fiction-motherhood-helen-phillips.html)).

This is why Hitchcock called it, "The Man Who Knew Too Much." Sometimes it's better not to know.

I just realized "expected type" as a trope of "what to expect when you're expecting"!

**Comment by dwijnand** [comment]
(By the way, I'm concerned that I might be coming across as combative towards you - I know you helped align the implementation with the spec and I'm sorry I didn't make my arguments at that time.  But I'm arguing against the specified behaviour, not against you or your efforts.)

**Comment by odersky** [comment]
I still think the spec and the current implementation is the best compromise. Examples:
```
// file1.scala
package p
class A

// file2.scala
package test
import p.A
object B extends A

//file3.scala
package test
class A
```
It would be very surprising if a random file in the same package (which might even be the empty package) overrides an explicitly given import. On the other hand:

```
// file4.scala
package test
import p._
class A
object B extends A
```
It should be blindingly obvious that we mean `p`'s A here, not the import. The general scope rules support that idea since they say that, within one scope, a definition shadows an import. Two definitions
in the same compilation unit can be constructed as being in the same "scope", whereas it's stretch to call two definitions in different files as being in the same scope. Technically this is true, but conceptually it's counter-intuitive.

**Comment by som-snytt** [comment]
(Also, I don't even care how anything works, in the sense that Scala has seen both good ideas and bad ones, and our collective job is to partition them through hard experience.)

**Comment by Jasper-M** [comment]
> The reason for the spec wording is that one often has little insight or control over what is in a package. E.g. anything could be dumped in the empty package. So it is dangerous and unintuitive that this should shadow an import.

I find this a pretty convincing argument. 
But in the following code it would seem safer to me if the compiler would tell me that there's a name clash, which I can fix, instead of just picking one over the other.

```
import java.util.Properties
class Properties
new Properties
```

I'm not sure how I feel about the wildcard case. Perhaps give a error for the previous code and a warning for the following:

```
import java.util._
class Properties
new Properties
```

**Comment by som-snytt** [comment]
I recently joked that, if challenged, Dr Odersky would say, "Not an authority," the way Janet the all-knowing assistant demurs on The Good Place by saying, "Not a robot," etc. I withdrew that joke, which was a good one, because Dr O popped up in the thread, like the scene from Annie Hall where Marshall McLuhan shows up.

I'd like to revise it now and suggest that Martin is not "the subject supposed to know" but the guy who should probably understand what we reached consensus about.

I tried and failed to google a quick Lacan link, but now I'd like to propose that Scaladoc be written in Lacanian prose, for example:

> Imagine you see in front of you a beautiful flower, or a ripe fruit. You reach out your hand to grab it. But at the moment you do, the flower, or the fruit, bursts into flames. In its place you see another hand appear, reaching back towards your own. (Seminar VIII, p.52; p.179).

https://www.lacanonline.com/2016/06/what-does-lacan-say-about-love/

That totally summarizes everything I've learned about functional programming. And the Scala community.

**Comment by som-snytt** [comment]
@Jasper-M - I think a lint option would help where confusion is possible.

**Comment by dwijnand** [comment]
Thanks, @odersky, for responding.

> It would be very surprising if a random file in the same package (which might even be the empty package) overrides an explicitly given import.

Seeing as this is compile-time, I think it's fine for the compiler to complain (either via a warning or an error) about such ambiguities.  Also given the JPMS.  I'm fine with doing something different about the empty package, as it's a weird one.

> It should be blindingly obvious that we mean `test`'s A here, not the import.

I disagree it's so obvious.

---

Continuing on my thought process that the compilation unit shouldn't matter, I explored below the various situations.

First given `file1.scala`:

```scala
package p { class A }
```

Here are the combinations, grouped up.  The idea is:

* `file2.scala` == line 2
* `file3.scala` == line 3
* `file4.scala` == line 2 + line 3
* `file2b.scala` == line 2b
* `file4b.scala` == line 2b + line 3

# named package

```scala
package test { import p.A; object B extends A } // line 2
package test { import p._; object B extends A } // line 2b (wildcard alternative)
package test { class A }                        // line 3
```

❌ error: ambiguous if `p.A` or `test.A`
🔧 fix by import renaming: `import test.{ A => TestA }` or `import p.{ A => PA }`

# empty package

```scala
package test { import p.A; object B extends A } // line 2
package test { import p._; object B extends A } // line 2b (wildcard alternative)
class A                                         // line 3
```

✅ works, uses `p.A`
⚠️ a lint warns that `<empty>.A` is shadowed by `p.A`; resolved by... deleting/moving `A`?  suppressing the warning at `import p.A`?

**Comment by odersky** [comment]
>> It should be blindingly obvious that we mean test's A here, not the import.

> I disagree it's so obvious.

I disagree. It _is_ obvious. It has been like this for all Scala versions and is completely in line with how Scala handles scopes. That part is non-negotiable.

**Comment by som-snytt** [comment]
We could add `-Xlint:captain-obvious` for the edge case that was corrected. I had a little trouble following Dale's example.
```
$ cat file1.scala file3.scala file4.scala 

package p { class A }

package test { class A { override def toString = "testy" } }

package test { import p._ ; class B extends A ; object Main extends App { println(new B()) } }
$ ~/scala-2.12.8/bin/scalac file1.scala file3.scala file4.scala && ~/scala-2.12.8/bin/scala test.Main
testy
$ scalac file1.scala file3.scala file4.scala && scala test.Main
test.B@4d1b0d2a
$ cat file1.scala file3.scala file4.scala > file5.scala
$ scalac file5.scala && scala test.Main
testy
```
It could be a somewhat expensive lint because it would inspect packages (and also package objects). It would warn for both cases: "Your import of test.A is shadowed by a definition in this compilation unit", and "The definition test.A in another compilation until is shadowed by your import."

**Comment by dwijnand** [comment]
> I disagree. It _is_ obvious. It has been like this for all Scala versions and is completely in line with how Scala handles scopes. That part is non-negotiable.

OK. Let me try again, holding this behaviour as non-negotiable:

```scala
package test { import p._; object B extends A }
package test { class A }
```

single file case: `object B` must be extending `test.A` (the import is unused)

```scala
package test { import p._; object B extends A }
```
```scala
package test { class A }
```

split file case: in order to not have different behaviours by splitting a file this must have the same behaviour: `object B` extends `test.A` (+ unused import)

Now, using named imports, not wildcards.

```scala
package test { import p.A; object B extends A }
package test { class A }
```

single file case: ambiguous reference error?

```scala
package test { import p.A; object B extends A }
```
```scala
package test { class A }
```

split file case: ambiguous reference error?

(Leaving the empty package case to one side for now.)

WDYT?

**Comment by som-snytt** [comment]
Today's Strands puzzle.

![Image](https://github.com/user-attachments/assets/e43bc762-d639-4180-bb21-971b92d9e4a5)
